compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs

```diff
@@ -53,9 +53,38 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
         let goals_certainty = self.try_evaluate_added_goals()?;
         let certainty = certainty.unify_with(goals_certainty);
 
-        let external_constraints = self.compute_external_query_constraints()?;
+        let response = match certainty {
+            Certainty::Yes | Certainty::Maybe(MaybeCause::Ambiguity) => {
+                let external_constraints = self.compute_external_query_constraints()?;
+                Response { var_values: self.var_values, external_constraints, certainty }
+            }
+            Certainty::Maybe(MaybeCause::Overflow) => {
+                // If we have overflow, it's probable that we're substituting a type
+                // into itself infinitely and any partial substitutions in the query
+                // response are probably not useful anyways, so just return an empty
+                // query response.
+                Response {
```

Comment:
can we instead add a `make_ambiguous_response_no_constraints` method to `eval_ctxt` and use that here (and in `fn flounder`)?

i feel like it would be nice to split the implementation of "make a response with no constraints from the actual logic.